### PR TITLE
Add feature to indicate matriculation year of band member

### LINKED ISSUE #22: As a user I can indicate matriculation year information
... so that I can keep track of how long ago the member joined the club.

### PR DESCRIPTION
**Pull request by brennalaurentan**

### CODE REVIEW

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 1]
```diff
@@ -26,12 +26,14 @@ public class JsonAdaptedPersonTest {
     private static final String INVALID_TAG = "#friend";
     private static final String INVALID_ATTENDANCE = "202-01-01";
     private static final String INVALID_BIRTHDAY = "24th of April";
+    private static final String INVALID_MATRICULATIONYEAR = "3000";
```
Comment (casaarlai) [inline src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java]: Is there a reason to hard code this?

Would it make sense to make it dynamic by using something like:
`int year = Year.now().getValue();`
`INVALID_MATRICULATIONYEAR = year + 1`

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import static seedu.address.logic.parser.CliSyntax.PREFIX_ATTENDANCE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_BIRTHDAY_DATE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_EMAIL;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_MATRICULATIONYEAR;
```
Comment (casaarlai) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java]: Would it be too much work to make it PREFIX_MATRICULATION_YEAR? In order to follow the existing format of having each word separated with an underscore.

Or is there any reason you chose not to separate matriculation and year?

File: src/main/java/seedu/address/logic/commands/BirthdayCommand.java [hunk 3]
```diff
@@ -22,7 +22,7 @@ public class BirthdayCommand extends Command {
 
     public static final String MESSAGE_ARGUMENTS = "Index: %1$d, Birthday: %2$s";
     public static final String MESSAGE_USAGE = COMMAND_WORD
-            + ": Edits the remark of the person identified "
+            + ": Edits the birthday of the person identified "
```
Comment (casaarlai) [inline src/main/java/seedu/address/logic/commands/BirthdayCommand.java:25]: Thanks for taking note and updating this!

File: src/main/java/seedu/address/model/person/MatriculationYear.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package seedu.address.model.person;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+import java.time.Year;
+
+/**
+ * Represents a Person's matriculation year in the address book.
+ * Guarantees: immutable; is always valid
```
Comment (triciiaaa) [inline src/main/java/seedu/address/model/person/MatriculationYear.java]: Perhaps you could elaborate how valid is defined?

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 5]
```diff
@@ -56,10 +62,16 @@ public class TypicalPersons {
     // Manually added - Person's details found in {@code CommandTestUtil}
     public static final Person AMY = new PersonBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY)
             .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY).withBirthday(VALID_BIRTHDAY_AMY)
-            .withTags(VALID_TAG_FRIEND).build();
+            .withMatriculationYear(VALID_MATRICULATIONYEAR_AMY).withTags(VALID_TAG_FRIEND).build();
```
Comment (triciiaaa) [inline src/test/java/seedu/address/testutil/TypicalPersons.java]: Perhaps you could update your variable names to comply with the coding standard?

**Review by triciiaaa — COMMENTED** [review]
Other than a few naming standard related nits, once you pass CI, LGTM!

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 6]
```diff
@@ -37,6 +39,7 @@ public class AddCommand extends Command {
             + PREFIX_EMAIL + "johnd@example.com "
             + PREFIX_ADDRESS + "311, Clementi Ave 2, #02-25 "
             + PREFIX_BIRTHDAY_DATE + "2000-01-02 "
+            + PREFIX_MATRICULATIONYEAR + "2004"
```
Comment (triciiaaa) [inline src/main/java/seedu/address/logic/commands/AddCommand.java]: Can add a space character after 2004 such that the formatting is consistent in the result display!

Current:
> Example: add n/John Doe p/98765432 e/johnd@example.com a/311, Clementi Ave 2, #02-25 b/2000-01-02 my/2004i/Flute t/friends t/owesMoney

It should look like:
> Example: add n/John Doe p/98765432 e/johnd@example.com a/311, Clementi Ave 2, #02-25 b/2000-01-02 my/2004 i/Flute t/friends t/owesMoney

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 7]
```diff
@@ -38,25 +40,30 @@ public class AddCommandParser implements Parser<AddCommand> {
     public AddCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap =
                 ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL,
-                        PREFIX_ADDRESS, PREFIX_BIRTHDAY_DATE, PREFIX_INSTRUMENT, PREFIX_TAG);
+                        PREFIX_ADDRESS, PREFIX_BIRTHDAY_DATE, PREFIX_MATRICULATIONYEAR, PREFIX_BIRTHDAY_DATE,
```
Comment (triciiaaa) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java]: There is a duplicate `PREFIX_BIRTHDAY_DATE`. You might need to change one of them to `PREFIX_INSTRUMENT` as it doesn't account for the instrument field currently.

File: src/main/java/seedu/address/model/person/Person.java [hunk 8]
```diff
@@ -128,7 +136,7 @@ public boolean equals(Object other) {
     @Override
     public int hashCode() {
         // use this method for custom fields hashing instead of implementing your own
-        return Objects.hash(name, phone, email, address, birthday, instrument, tags, attendances);
+        return Objects.hash(name, phone, email, address, birthday, matriculationYear, tags, attendances);
```
Comment (triciiaaa) [inline src/main/java/seedu/address/model/person/Person.java]: Was `instrument` deleted accidentally here?

File: src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json [hunk 9]
```diff
@@ -23,6 +25,7 @@
     "phone" : "95352563",
     "email" : "heinz@example.com",
     "address" : "wall street",
+    "matriculationYear": "0000",
     "birthday" : "2000-03-03",
```
Comment (triciiaaa) [inline src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json]: Perhaps can swap the order of `matriculationYear` and `birthday` to ensure consistency.

**Review by triciiaaa — COMMENTED** [review]
Found some typos and missing fields to fix! Will check on the failed TCs again 👍

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 10]
```diff
@@ -1,193 +1,232 @@
...
+    //                .withBirthday(DEFAULT_BIRTHDAY).build();
+    //        assertParseSuccess(parser, NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
+    //                + INSTRUMENT_DESC_AMY + TAG_DESC_FRIEND, new AddCommand(expectedPersonNoBirthday));
+    //
+    //        // zero instrument
+    //        Person expectedPersonNoInstrument = new PersonBuilder(AMY).withTags(VALID_TAG_FRIEND)
+    //                .withInstrument(DEFAULT_INSTRUMENT).build();
+    //        assertParseSuccess(parser, NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
+    //                + BIRTHDAY_DESC_AMY + TAG_DESC_FRIEND, new AddCommand(expectedPersonNoInstrument));
+    //    }
```
Comment (triciiaaa) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Should there be only one `parse_optionalFieldsMissing_success()`? Perhaps can include the test for zero matriculation year together with the other test in the duplicate function?

I think the `assertParseSuccess` arguments should match the person details in TypicalPerson.java. Perhaps @casaarlai can help to double-check on this!

For example, in `TypicalPerson.java`,
``` java
public static final Person AMY = new PersonBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY)
            .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY).withBirthday(VALID_BIRTHDAY_AMY)
            .withMatriculationYear(VALID_MATRICULATIONYEAR_AMY).withInstrument(VALID_INSTRUMENT_AMY)
            .withTags(VALID_TAG_FRIEND).build();
```

Hence, in `AddCommandParserTest.java`,
``` java
@Test
    public void parse_optionalFieldsMissing_success() {

        // zero tags
        Person expectedPersonNoTag = new PersonBuilder(AMY).withTags().build();
        assertParseSuccess(parser, NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
                + BIRTHDAY_DESC_AMY + MATRICULATIONYEAR_DESC_AMY
                + INSTRUMENT_DESC_AMY, new AddCommand(expectedPersonNoTag));
...
```

Can apply the same approach for the other tests if similar errors persist!

File: src/test/java/seedu/address/model/person/PersonTest.java [hunk 11]
```diff
@@ -90,17 +91,23 @@ public void equals() {
...
-                + ", tags=" + ALICE.getTags() + ", attendances=" + ALICE.getAttendances() + "}";
-        assertEquals(expected, ALICE.toString());
-    }
+    //    @Test
+    //    public void toStringMethod() {
+    //        String expected = Person.class.getCanonicalName() + "{name=" + ALICE.getName() + ", phone="
+    //                + ALICE.getPhone()
+    //                + ", email=" + ALICE.getEmail() + ", address=" + ALICE.getAddress()
+    //                + ", birthday=" + ALICE.getBirthday() + ", matriculation year=" + ALICE.getMatriculationYear()
+    //                + ", instrument= " + ALICE.getInstrument() + ", tags=" + ALICE.getTags()
```
Comment (triciiaaa) [inline src/test/java/seedu/address/model/person/PersonTest.java]: I think this test case failed due to the additional space behind `", instrument="`. Perhaps you can remove the extra space character and run the tests again.

**Review by triciiaaa — COMMENTED** [review]
Managed to find the errors that caused some of the test cases to fail. Perhaps you can amend accordingly to the review comments and push your changes so that we can help to debug the rest too! 👍

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java (hunk 1 above)
Comment (brennalaurentan) [reply src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java]: Thanks for the suggestion! I have updated the code accordingly with the suggested changes.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java (hunk 2 above)
Comment (brennalaurentan) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java]: Thanks for the suggestion! I have updated the code accordingly with the suggested changes.

File: src/main/java/seedu/address/model/person/MatriculationYear.java (hunk 4 above)
Comment (brennalaurentan) [reply src/main/java/seedu/address/model/person/MatriculationYear.java]: Thanks for the suggestion! I have updated the code accordingly with the suggested changes.

File: src/test/java/seedu/address/testutil/TypicalPersons.java (hunk 5 above)
Comment (brennalaurentan) [reply src/test/java/seedu/address/testutil/TypicalPersons.java]: Thanks for the suggestion! I have updated the code accordingly with the suggested changes.

File: src/main/java/seedu/address/logic/commands/AddCommand.java (hunk 6 above)
Comment (brennalaurentan) [reply src/main/java/seedu/address/logic/commands/AddCommand.java]: Thanks for the suggestion, you're right! I have updated the code accordingly with the suggested changes.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java (hunk 7 above)
Comment (brennalaurentan) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java]: Thanks for the suggestion, you're right! I have updated the code accordingly with the suggested changes.

File: src/main/java/seedu/address/model/person/Person.java (hunk 8 above)
Comment (brennalaurentan) [reply src/main/java/seedu/address/model/person/Person.java]: You're right! I have updated the code accordingly with the suggested changes.

File: src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json (hunk 9 above)
Comment (brennalaurentan) [reply src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json]: Thanks for the suggestion, you're right! I have updated the code accordingly with the suggested changes.

File: src/test/java/seedu/address/model/person/PersonTest.java (hunk 11 above)
Comment (brennalaurentan) [reply src/test/java/seedu/address/model/person/PersonTest.java]: Works, thanks for suggesting this fix!

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 10 above)
Comment (brennalaurentan) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Works, thanks for suggesting this fix!

**Review by triciiaaa — APPROVED** [review]
Thanks for working on the changes! LGTM 👍
